### ISSUE TITLE
file_get_contents incorrectly flagged as bad behaviour

### ISSUE THREAD

**Issue by diddledan** [WordPress-Coding-Standards/WordPress-Coding-Standards#943]
`file_get_contents()` is being flagged as bad with the suggested replacement being `wp_remote_get()`. This is wrong for local files, because `wp_remote_get()` does not load files from the filesystem (unless I'm mistaken?)

**Comment by JDGrimes** [comment]
It appears that this is coming from the [`WordPress.WP.AlternativeFunctions` sniff](https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/blob/c0d7578a80314ddbaf9d188553c1366d63f5180a/WordPress/Sniffs/WP/AlternativeFunctionsSniff.php#L58-L64). This check was introduced in #633 (See https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/pull/633#pullrequestreview-12368901).

I guess @grappler introduced this because sometimes `file_get_contents()` is used to retrieve remote URLs. The fact that it doesn't always apply is probably why it was a warning and not an error.

Maybe this check should be removed, since there will still be an error given for `file_get_contents()`:

```
 3 | WARNING | File operations should use WP_Filesystem methods instead
   |         | of direct PHP filesystem calls. Found: file_get_contents()
   |         | (WordPress.WP.AlternativeFunctions.file_system_read_file_get_contents)
```

That error could be updated to also mention the possibility that it is being used for a URL rather than to access the filesystem.

Or perhaps it would be possible to detect whether a URL was being passed in some cases, and only give the warning then. It wouldn't always be able to detect that though.

**Comment by grappler** [comment]
I think the easiest solution would be to improve the message to something like:
> Use `wp_remote_get()` for remote URLs and WP_Filesystem methods for local files.

**Comment by aristath** [comment]
I don't think that `file_get_contents` should be flagged if not using a URL...
I can think of some valid use-cases that would require it (for example getting the contents of a JSON file) and the `$wp_filesystem` won't work in case the user is using `ftpext` and has not defined their FTP credentials in their `wp-config.php`.
It may be a file operation, but it's not like `file_put_contents`...

**Comment by JDGrimes** [comment]
> It may be a file operation, but it's not like file_put_contents...

I think that is actually a good point. The main reason for using `$wp_filesystem` is so that the correct permissions are set on write, if I'm not mistaken. So when reading it isn't as important, except in cases when you can only read with `ftpext` or something, if that is ever true.

**Comment by paulcanning** [comment]
How do I tell Codesniffer to ignore this sniff, as I am solely using it for local file contents.

**Comment by jrfnl** [comment]
@paulcanning Generally speaking, you're normally better off not ignoring the sniff completely, but selectively ignoring specific error messages for specific files or - depending on the sniff - there may be properties you can set to customize the behaviour.

Taking the specific message mentioned above. you can [selectively ignore](https://github.com/squizlabs/PHP_CodeSniffer/wiki/Annotated-ruleset.xml) it by adding something along the lines of the following to your custom ruleset:
```xml
<rule ref="WordPress.WP.AlternativeFunctions.file_system_read_file_get_contents">
    <exclude-pattern>/path/to/the/file\.php</exclude-pattern>
</rule>
```

Or if you want to ignore the whole `file_system_read` group for your complete project, you can do so om your ruleset by [using a custom property](https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/wiki/Customizable-sniff-properties#excluding-a-group-of-checks):
```xml
<rule ref="WordPress.WP.AlternativeFunctions">
	<properties>
		<property name="exclude" value="file_system_read" />
	</properties>
</rule>
```

**Comment by jrfnl** [comment]
FYI: ignoring this error should, in most cases, no longer be necessary. PR #1374 which has just been merged adds a check which tries to determine if a local file or remote file is being accessed and the sniff will now only throw the `warning` if a remote file is being requested or if local/remote could not be determined.

**Comment by mckravchyk** [comment]
@jrfnl I still get that warning on something like: `file_get_contents(dirname(__FILE__) . '/assets/icons/collapse.svg')`

**Comment by jrfnl** [comment]
@mckravchyk Please open a new issue with all the details and we'll try and have a look.
